File: compare_strings.py
from typing import Tuple, Optional, TextIO

from regions import Region
from top import Top

import sys
import argparse
import time

ALPHABET = {'A', 'C', 'G', 'N', 'T'}

def ERROR(title: str, msg : str):
    ORANGE = "\033[33m"
    RED = "\033[31m"
    BOLD = "\033[1m"
    RESET = "\033[0m"
    print(f"{BOLD}({ORANGE}{title}{RESET}) {BOLD}{RED}[ERREUR]{RESET} {msg}")
    
def WARNING(title: str, msg : str):
    ORANGE = "\033[33m"
    RED = "\033[31m"
    BOLD = "\033[1m"
    RESET = "\033[0m"
    print(f"{BOLD}({RED}{title}{RESET}) {BOLD}{ORANGE}[WARNING]{RESET} {msg}")

def fasta_open_and_read_header(path: str) -> Tuple[Optional[TextIO], Optional[str]]:
    try:
        fp = open(path, "r")
    except OSError as error:
        ERROR("ULB", f"Fichier ou dossier inéxistant \033[33m-->\033[0m \033[31m{error}\033[0m")
        return None, None

    header = fp.readline()
    if header == "":
        fp.close()
        ERROR("ULB", f"Fichier Vide \033[33m-->\033[0m Il doit contenir : \033[31mLigne 1\033[0m : Header et \033[31mLigne 2\033[0m : L'ADN")
        return None, None

    header = header.strip()
    if not header.startswith(">"):
        fp.close()
        ERROR("ULB", f"Nom de séquence incorrect \033[33m-->\033[0m Il doit étre de cette forme : \033[31mLigne 1\033[0m : >chr1")
        return None, None

    name = header[1:]
    
    return fp, name


def fasta_read_bases(fp: Optional[TextIO], n: int) -> Optional[str]:
    if fp is None:
        ERROR("ULB", f"ERREUR : lecture ADN impossible, fichier non ouvert dans fasta_open_and_read_header().")
        return None

    bases = []
    
    while len(bases) < n:
        c = fp.read(1)

        if c == "":
            break

        if c == "\n" or c == "\r":
            continue

        if c == ">":
            ERROR("ULB", f"ERREUR : séquence ADN incorrecte, '>' trouvé dans la ligne de séquence.")
            return None

        bases.append(c)

    return "".join(bases)


def csv_open(path: str) -> Optional[TextIO]:
    try:
        return open(path, "w")
    except OSError as error:
        ERROR("ULB", f"Fichier ou dossier inéxistant \033[33m-->\033[0m \033[31m{error}\033[0m")
        return None


def csv_write_mutation(fp: TextIO, name: str, pos: int, ref: str, alt: str) -> None:
    fp.write(f"{name},{pos},{ref},{alt}\n")

"""
    J'ai utiliser l'ia uniquement pour cette fonction car c'était à la fin de mon projet que je l'es remarqué
    aussi dans le c++ la fonction arrondi_2
"""
def arrondi_2(x):
    return f"{x + 1e-12:.2f}"

def csv_write_regions(fp: TextIO, len_sequence : int, name: str, top_regions : list) -> None:
    for region in top_regions:
        b = []
        for i in region.statistique():
            b.append(arrondi_2(i))
        fraction_a_final = ",".join(b)
        fraction_t_final = arrondi_2(region.fraction_total(len_sequence))
        fp.write(f"{name},{region.start},{region.end()},{region.len_r},{fraction_t_final},{fraction_a_final}\n")

def csv_close(fp: Optional[TextIO]) -> None:
    if fp is not None:
        fp.close()
        
def check_ALPHABET(sequence): 
    for i in sequence:
        if i not in ALPHABET:
            return False
    return True

def compare_sequence(sequence_ref, sequence, name_seq, path_out, len_seq, top_regions, region):
    for pos in range(0, len(sequence_ref)):
        if sequence_ref[pos] != sequence[pos]:
            if region != None:
                top_regions.update_top(region)
                region = None
            csv_write_mutation(path_out, name_seq, len_seq + pos, sequence_ref[pos], sequence[pos])
        else:
            if region == None:
                region = Region(len_seq + pos, sequence_ref[pos])
            else:
                region.update_compteur(sequence_ref[pos])
    return region

def main_read_buffer(name_seq, fp_ref, fp_seq, fp_out, buffer_size, fp_out_regs):
    seq_ref, seq = [], []
    len_seq = 0
    
    top_regions = Top()
    region = None
        
    while not seq_ref == "" and not seq == "":
        seq_ref = fasta_read_bases(fp_ref, buffer_size)
        seq = fasta_read_bases(fp_seq, buffer_size)
        
        if (seq_ref is None) or (seq is None):
            ERROR("ULB", "ERREUR : une erreur c'est produite lors de la lecture des séquences.")
            return 1
        
        if seq_ref == "" and seq == "":
            break
        
        if seq_ref == "" and not seq == "":
            ERROR("ULB", f"ERREUR : les tailles de séquence d'ADN sont différent, ils doivent étre égaux")
            return 1
        
        if not seq_ref == "" and seq == "":
            ERROR("ULB", f"ERREUR : les tailles de séquence d'ADN sont différent, ils doivent étre égaux")
            return 1
    
        if len(seq_ref) != len(seq):
            ERROR("ULB", f"ERREUR : les tailles de séquence d'ADN sont différent, ils doivent étre égaux")
            return 1
            
        if not check_ALPHABET(seq_ref) or not check_ALPHABET(seq):
            ERROR("ULB", "ERREUR : une des sequences contient un caractere qui n'appartient pas a l'alphabet ACGNT.")
            return 1
            
        size = len(seq_ref) % buffer_size
        region = compare_sequence(seq_ref, seq, name_seq, fp_out, len_seq, top_regions, region)
        
        if size > 0:
            len_seq += size
        else:
            len_seq += buffer_size
    
    if region != None:
        top_regions.update_top(region) 
    
    regions_top = top_regions.get_regions()
    csv_write_regions(fp_out_regs, len_seq, name_seq, regions_top)
        
    csv_close(fp_out)
    csv_close(fp_out_regs)
    fp_ref.close()
    fp_seq.close()


def main():
    start = time.time()
    parser = argparse.ArgumentParser()
    parser.add_argument("path_seq_ref", type=str, help="Entrée de la premiére séquence de ref à comparer (FASTA).")
    parser.add_argument("path_seq", type=str, help="Entrée de la deuxiéme séquence à comparer (FASTA).")
    parser.add_argument("path_output_sub", type=str, help="Sorti du résultat des substitues (CSV).")
    parser.add_argument("path_output_regs", type=str, help="Sorti du résultat des régions (CSV).")
    parser.add_argument("buffer_size", type=int, help="Taille de lecture de caractére.")
    args = parser.parse_args()
    
    path_ref = args.path_seq_ref
    path_seq = args.path_seq
    path_out = args.path_output_sub
    path_out_regs = args.path_output_regs
    buffer_size = args.buffer_size
    
    if buffer_size <= 0:
        ERROR("ULB", "buffer_size doit être strictement positif")
        return 1
        

    fp_ref, name_ref = fasta_open_and_read_header(path_ref)
    if fp_ref is None:
        return 1

    fp_seq, name_seq = fasta_open_and_read_header(path_seq)
    if fp_seq is None:
        fp_ref.close()
        return 1

    if name_ref != name_seq:
        ERROR("ULB", f"Nom de séquence de référence : \033[33m{name_ref}\033[0m différent que le nom de séquence à comparer \033[33m{name_seq}\033[0m")
        fp_ref.close()
        fp_seq.close()
        return 1

    fp_out = csv_open(path_out)
    if fp_out is None:
        fp_ref.close()
        fp_seq.close()
        return 1

    fp_out_regs = csv_open(path_out_regs)
    if fp_out_regs is None:
        csv_close(fp_out)
        fp_ref.close()
        fp_seq.close()
        return 1

    read = main_read_buffer(name_ref, fp_ref, fp_seq, fp_out, buffer_size, fp_out_regs)
    if read == 1:
        csv_close(fp_out)
        csv_close(fp_out_regs)
        fp_ref.close()
        fp_seq.close()
        return 1
                
    end = time.time()
    print(f"Temps d'exécution : {(end - start)} secondes")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: compare_strings.cpp
#include <cstdio>
#include <cstdlib>

#include "utils_fasta.h"
#include "utils_csv.h"

#include "region.hpp"
#include "top.hpp"

#include <iostream>
#include <cstring>
#include <cstddef>
#include <time.h>

using namespace std;

static const size_t MAX_SEQ_NAME_LEN = 101; // 100 caractères + '\0'

void closeAllFile(FILE* fp_out, FILE* fp_out_reg, FILE* fp_ref, FILE* fp_seq) {
    csv_close(fp_out);
    csv_close(fp_out_reg);
    fasta_close(fp_ref);
    fasta_close(fp_seq);
}

bool check_ALPHABET(const char* seq, size_t len) {
    for (size_t i = 0; i < len; i++) {
        if (seq[i] != 'A' && seq[i] != 'C' && seq[i] != 'G' && seq[i] != 'N' && seq[i] != 'T') {
            return false;   
        }
    }
    return true;
}

int compare_sequence(const char* seq_ref, const char* seq, char* name_ref, FILE* fp_out, size_t len_all_seq, size_t len_seq, Top& top, Region*& region) {
    for (size_t i = 0; i < len_seq; i++) {
        if (seq_ref[i] != seq[i]) {
            if (region != nullptr) {
                top.update_top(*region);
                delete region;
                region = nullptr;
            }
            CsvStatus cs = csv_write_mutation(fp_out, name_ref, len_all_seq + i, seq_ref[i], seq[i]);
            if (cs != CSV_OK) {
                cout << "ERREUR : impossible d'ecrire dans le fichier CSV des substitutions." << endl;
                return -1;
            }
        } else {
            if (region == nullptr) {
                region = new Region(len_all_seq + i, seq_ref[i]);
            } else {
                region->update_compteur(seq_ref[i]);
            }
        }
    }
    return 1;
}

int read_error(FastaStatus fs, const char* file) {
    switch (fs) {
        case FASTA_ERR_OPEN:
            cout << "ERREUR " << (file) << " : ouverture du fichier FASTA impossible." << endl; 
            break;
        case FASTA_ERR_EMPTY:
            cout << "ERREUR " << (file) << " : le fichier FASTA ne contient aucune sequence." << endl; 
            break;
        case FASTA_ERR_FORMAT:
            cout << "ERREUR " << (file) << " : le fichier n'est pas au format FASTA valide (en-tete manquant, plusieurs sequences invalide, ...) ." << endl; 
            break;
        case FASTA_ERR_NAME_TOO_LONG:
            cout << "ERREUR " << (file) << " : le nom de la sequence depasse la taille autorisee." << endl; 
            break;
        case FASTA_ERR_IO:
            cout << "ERREUR " << (file) << " : une erreur d'entree/sortie a été repéré lors de la lecture." << endl; 
            break;
        default:
        cout << "ERREUR " << (file) << " : probléme non-identifié."<< endl; 
            break;
        }
    return -1;
}

int main_read_buffer(char* name_seq, FILE* fp_ref, FILE* fp_seq, FILE* fp_out, size_t buffer_size, FILE* fp_out_regions) {
    char* seq_ref = new char[buffer_size];
    char* seq= new char[buffer_size];
    size_t n_ref = 0;
    size_t n_seq = 0;
    size_t len_all_sequence = 0;

    Top top_regions = Top();
    Region* region = nullptr;

    FastaStatus fs;

    while (true) {
        fs = fasta_read_bases(fp_ref, seq_ref, buffer_size, &n_ref);
        if (fs != FASTA_OK) {
            const char* file = "séquence de référence";
            int error = read_error(fs, file);
            if (error == -1) {
                delete region;
                delete[] seq_ref;
                delete[] seq;
                closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
                return -1;
            }
        }

        fs = fasta_read_bases(fp_seq, seq, buffer_size, &n_seq);
        if (fs != FASTA_OK) {
            const char* file = "séquence à comparer";
            int error = read_error(fs, file);
            if (error == -1) {
                delete region;
                delete[] seq_ref;
                delete[] seq;
                closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
                return -1;
            }
        }

        if (n_ref == 0 and n_seq == 0) {
            break;
        } 

        if (n_ref != n_seq) {
            cout << "ERREUR : les tailles de séquence d'ADN sont différent, ils doivent étre égaux." << endl;
            delete region;
            delete[] seq_ref;
            delete[] seq;
            closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
            return -1;
        }

        if (not check_ALPHABET(seq_ref, n_ref) or not check_ALPHABET(seq, n_seq)) {
            cout << "ERREUR : une des sequences contient un caractere qui n'appartient pas a l'alphabet ACGNT." << endl;
            delete region;
            delete[] seq_ref;
            delete[] seq;
            closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
            return -1;
        }

        size_t size = n_ref % buffer_size;

        int ok = compare_sequence(seq_ref, seq, name_seq, fp_out, len_all_sequence, n_ref, top_regions, region);

        if (ok == -1) {
            delete region;
            delete[] seq_ref;
            delete[] seq;
            closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
            return -1;
        }

        if (size > 0) {
            len_all_sequence += size;
        } else {
            len_all_sequence += buffer_size;
        }
    }

    if (region != nullptr) {
        top_regions.update_top(*region);
    }

    CsvStatus cs = csv_write_region(fp_out_regions, len_all_sequence, name_seq,top_regions.get_first(),top_regions.get_second(),top_regions.get_third());

    if (cs != CSV_OK) {
        cout << "ERREUR : impossible d'ecrire dans le fichier CSV des regions." << endl;
        delete region;
        delete[] seq_ref;
        delete[] seq;
        closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
        return -1;
    }

    delete region;
    delete[] seq_ref;
    delete[] seq;
    closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
    return 1;
}

int main(int argc, char* argv[]) {
    clock_t start = clock();

    if (argc != 6) {
        cout << "Argument #" << argc << " manquant dans l'éxécution, 'chemin_ref_seq' 'chemin_seq' 'chemin_output_subs' 'chemin_output_regs' 'buffer_size'" << endl; 
        return -1;
    }

    const char* path_ref = argv[1];
    const char* path_seq = argv[2];
    const char* path_out = argv[3];
    const char* path_out_regs = argv[4];
    const char* arg_bufer_size = argv[5];
    
    int buffer_size_int = atoi(arg_bufer_size);
    if (buffer_size_int <= 0) {
        cout << "ERREUR : buffer_size doit étre strictement positif" << endl;
        return -1;
    }
    size_t buffer_size = static_cast<size_t>(buffer_size_int);

    FILE* fp_ref = nullptr;
    FILE* fp_seq = nullptr;

    char name_ref[MAX_SEQ_NAME_LEN];
    char name_seq[MAX_SEQ_NAME_LEN];

    FastaStatus fs;

    fs = fasta_open_and_read_header(path_ref, &fp_ref, name_ref, MAX_SEQ_NAME_LEN);
    if (fs != FASTA_OK) {
        const char* file = "séquence de référence";
        int error = read_error(fs, file);
        if (error == -1) {
            fasta_close(fp_ref);
            fasta_close(fp_seq);
            return -1;
        }
    }


    fs = fasta_open_and_read_header(path_seq, &fp_seq, name_seq, MAX_SEQ_NAME_LEN);
    if (fs != FASTA_OK) {
        const char* file = "séquence à comparer";
        int error = read_error(fs, file);
        if (error == -1) {
            fasta_close(fp_ref);
            fasta_close(fp_seq);
            return -1;
        }
    }

    if (strcmp(name_ref, name_seq) != 0) {
        cout << "ERREUR (Nom séquence) : Nom de séquence différent que le nom de séquence a comparer" << endl; 
        fasta_close(fp_ref);
        fasta_close(fp_seq);
        return -1;
    }


    FILE* fp_out = nullptr;
    FILE* fp_out_regions = nullptr;

    CsvStatus cs;
    cs = csv_open(path_out, &fp_out);
    if (cs != CSV_OK) {
        switch (cs) {
            case CSV_ERR_OPEN:
                cout << "ERREUR (file csv) : ouverture du fichier CSV impossible." << endl; 
                break;
            case CSV_ERR_IO:
                cout << "ERREUR (file csv) : une erreur d'entree/sortie a été repéré lors de l'écriture."<< endl; 
                break;
            default:
                cout << "ERREUR (file csv) : probléme non-identifié." << endl; 
                break;
        }
        closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
        return -1;
    }


    cs = csv_open(path_out_regs, &fp_out_regions);
    if (cs != CSV_OK) {
        switch (cs) {
            case CSV_ERR_OPEN:
                cout << "ERREUR (file csv) : ouverture du fichier CSV impossible." << endl; 
                break;
            case CSV_ERR_IO:
                cout << "ERREUR (file csv) : une erreur d'entree/sortie a été repéré lors de l'écriture."<< endl; 
                break;
            default:
                cout << "ERREUR (file csv) : probléme non-identifié." << endl; 
                break;
        }
        closeAllFile(fp_out, fp_out_regions, fp_ref, fp_seq);
        return -1;
    }


    int finish = main_read_buffer(name_seq, fp_ref, fp_seq, fp_out, buffer_size, fp_out_regions);
    if (finish == -1) {
        return -1;
    }
    clock_t end = clock();
    cout << "Temps d'execution : " << double(end - start) / CLOCKS_PER_SEC << " secondes" << endl;
    return 0;
}
